remove simple .data from torch/nn

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31482 remove simple .data from torch/nn**
* #31481 remove .data from torch/nn doc
* #31480 remove .data from torch/jit
* #31479 Remove some .data and clearly mark why we keep some

Differential Revision: [D19303243](https://our.internmc.facebook.com/intern/diff/D19303243)

albanD committed:
remove simple .data from torch/nn

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 26ab08e:


* **2/2** broken upstream at merge base fc3103b (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=fc3103b1162a3312e4ad74005d0b698a333c0750&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=26ab08e82e35f0ceff2ca811282f31eefb56eed2).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_6_gcc5_4_test](https://dr.pytorch.org/build-details.html?build_id=77618179)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

albanD committed:
Update on "remove simple .data from torch/nn"




[ghstack-poisoned]

albanD committed:
Update on "remove simple .data from torch/nn"




[ghstack-poisoned]

albanD committed:
Update on "remove simple .data from torch/nn"




[ghstack-poisoned]

albanD committed:
Update on "remove simple .data from torch/nn"




[ghstack-poisoned]

albanD committed:
Update on "remove simple .data from torch/nn"




[ghstack-poisoned]

albanD committed:
Update on "remove simple .data from torch/nn"


Differential Revision: [D19303243](https://our.internmc.facebook.com/intern/diff/D19303243)

[ghstack-poisoned]

albanD commented:
Error from the bot, this PR was not merged in the given commit

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -763,11 +763,9 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                     continue
 
-                if isinstance(input_param, Parameter):
-                    # backwards compatibility for serialized parameters
```
do you know what "backwards compatibility for serialized parameters" means here?

albanD replied:
This comes from a long long time ago [here](https://github.com/pytorch/pytorch/pull/451/files#diff-527b04fec26e044c060ad28b22e842beR240-R242). We used to store `Parameters` in the state_dict. Then we stopped.
Now that we use `no_grad()`. We don't care if it is a Parameter or not. So this is not needed anymore.

gchanan replied:
ah I see, it's just to get the tensor out from the Parameter.

albanD replied:
Yes.

gchanan commented on torch/nn/modules/module.py:
```diff
@@ -763,11 +763,9 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
-                    input_param = input_param.data
                 try:
-                    param.copy_(input_param)
+                    with torch.no_grad():
```
so if I'm reading this correctly, previously we didn't track autograd history through Parameters, but we did through Tensors?  And now we aren't tracking autograd history through either?  Why?

albanD replied:
`param` here is the result of `v.data` above. So changes to it are never tracked.
The new implementation makes that clearer that we never track history and just change the content in a non-differentiable manner.
